packages/flutter_tools/lib/src/device.dart

```diff
@@ -32,11 +32,28 @@ class DeviceManager {
 
   final List<DeviceDiscovery> _deviceDiscoverers = <DeviceDiscovery>[];
 
+  String _specifiedDeviceId;
+
   /// A user-specified device ID.
-  String specifiedDeviceId;
+  String get specifiedDeviceId {
+    if (_specifiedDeviceId == null || _specifiedDeviceId == 'all')
+      return null;
+    return _specifiedDeviceId;
+  }
+
+  set specifiedDeviceId(String id) {
+    _specifiedDeviceId = id;
+  }
 
+  /// True when the user has specified a single specific device.
   bool get hasSpecifiedDeviceId => specifiedDeviceId != null;
 
+  /// True when the user has specified all devices by setting
+  /// specifiedDeviceId = 'all'.
+  bool get hasSpecifiedAllDevices {
+    return _specifiedDeviceId != null && _specifiedDeviceId == 'all';
```

Comment:
`--use-application-binary` and `-d all` will only work if the application binary is compatible with all attached devices. A mixed environment (iOS/Android) wouldn't work.